File: 106-construct-binary-tree-from-inorder-and-postorder-traversal/construct-binary-tree-from-inorder-and-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int postOrderIndex;
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        if (inorder.length == 0) return null;
        postOrderIndex = postorder.length - 1;
        return buildTree(inorder, postorder, 0, inorder.length - 1);
    }
    public TreeNode buildTree(int[] inorder, int[] postOrder, int inStart, int inEnd) {
        if (inStart > inEnd) return null;
        TreeNode newNode = new TreeNode(postOrder[postOrderIndex--]);
        if (inStart == inEnd) {
            return newNode;
        }
        int index = search(inorder, newNode.val, inStart, inEnd);
        newNode.right = buildTree(inorder, postOrder, index+1,inEnd);
        newNode.left = buildTree(inorder, postOrder, inStart, index-1);
        return newNode;
    }
    public int search(int[] in, int val, int start, int end) {
        for (int i=start; i <= end; i++) {
            if (in[i] == val) return i;
        }
        return -1;
    }
}

File: 106-construct-binary-tree-from-inorder-and-postorder-traversal/construct-binary-tree-from-inorder-and-postorder-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, inorder: List[int], postorder: List[int]) -> Optional[TreeNode]:
        post_index = len(postorder) - 1
        def build_tree(current_inorder):
            nonlocal post_index
            if postorder == -1 or len(current_inorder) == 0:
                return None
            curr_val = postorder[post_index]
            post_index -= 1
            current_root = TreeNode(curr_val)
            curr_index = current_inorder.index(curr_val)
            left_inorder = current_inorder[:curr_index]
            right_inorder = current_inorder[curr_index + 1:]
            current_root.right = build_tree(right_inorder)
            current_root.left = build_tree(left_inorder)
            return current_root
        return build_tree(inorder)
